dbt/models/reporting/reporting.ratio_stats.py

```diff
@@ -162,120 +134,62 @@ def report_summarise(df, geography_id, geography_type):
         "sale_year",
     ]
 
-    df["geography_id"] = df[geography_id].astype(str)
-    df["geography_type"] = geography_type
-
-    # Remove groups with less than three observations
-    # TODO: Remove/upgrade detect_chasing output
-    df["n"] = df.groupby(group_cols)["ratio"].transform("count")
-    df = df[df["n"] > 3]
-    df = df.groupby(group_cols).apply(
-        lambda x: pd.Series(
-            {
-                "sale_n": np.size(x["triad"]),
-                "ratio": ccao_median(x["ratio"]),
-                "cod": ccao_cod(ratio=x["ratio"]),
-                "mki": ccao_mki(fmv=x["fmv"], sale_price=x["sale_price"]),
-                "prd": ccao_prd(fmv=x["fmv"], sale_price=x["sale_price"]),
-                "prb": ccao_prb(fmv=x["fmv"], sale_price=x["sale_price"]),
-                "detect_chasing": False,
-                "within_20_pct": sum(abs(1 - x["ratio"]) <= 0.20),
-                "within_10_pct": sum(abs(1 - x["ratio"]) <= 0.10),
-                "within_05_pct": sum(abs(1 - x["ratio"]) <= 0.05),
-            }
+    out = (
+        df.withColumn("geography_id", col(geography_id).cast("string"))
+        .withColumn("geography_type", lit(geography_type))
+        .groupby(group_cols)
+        .applyInPandas(
```

Comment:
The key to making things faster was just using native PySpark abstractions. Here `df` is a Spark table, and `applyInPandas` does all the heavy lifting of running a lambda function across all the specified grouping columns.